src/VisualStudio/Core/Def/Packaging/PackageSearchService.cs

```diff
@@ -0,0 +1,752 @@
+                {
+                    return _database_doNotAccessDirectly;
+                }
+            }
+
+            set
+            {
+                lock (_gate)
+                {
+                    _database_doNotAccessDirectly = value;
+                }
+            }
+        }
+
+        private void LogInfo(string text) => _logService.LogInfo(text);
+
+        private void LogException(Exception e, string text) => _logService.LogException(e, text);
+
+        /// <summary>
+        /// Internal for testing purposes.
+        /// </summary>
+        internal async Task UpdateDatabaseInBackgroundAsync()
+        {
+            // Keep on looping until we're told to shut down.
+            while (!_cancellationToken.IsCancellationRequested)
+            {
+                LogInfo("Starting update");
+                try
+                {
+                    var delayUntilNextUpdate = await UpdateDatabaseInBackgroundWorkerAsync().ConfigureAwait(false);
+
+                    LogInfo($"Waiting {delayUntilNextUpdate} until next update");
+                    await Task.Delay(delayUntilNextUpdate, _cancellationToken).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    LogInfo("Update canceled. Ending update loop");
+                    return;
+                }
+            }
+        }
+
+        /// <returns>The timespan the caller should wait until calling this method again.</returns>
+        private async Task<TimeSpan> UpdateDatabaseInBackgroundWorkerAsync()
+        {
+            // Attempt to update the local db if we have one, or download a full db
+            // if we don't.  In the event of any error back off a minute and try 
+            // again.  Lot of errors are possible here as IO/network/other-libraries
+            // are involved.  For example, we might get errors trying to write to 
+            // disk.
+            try
+            {
+                CleanCacheDirectory();
+
+                // If we have a local database, then see if it needs to be patched.
+                // Otherwise download the full database.
+                //
+                // (intentionally not wrapped in IOUtilities.  If this throws we want to restart).
+                if (_ioService.Exists(_databaseFileInfo))
+                {
+                    LogInfo("Local database file exists. Patching local database");
+                    return await PatchLocalDatabaseAsync().ConfigureAwait(false);
+                }
+                else
+                {
+                    LogInfo("Local database file does not exist. Downloading full database");
+                    return await DownloadFullDatabaseAsync().ConfigureAwait(false);
+                }
+            }
+            catch (OperationCanceledException)
+            {
+                // Just allow our caller to handle this (they will use this to stop their loop).
```

Comment:
I personally far prefer the code that separates out the different catch clauses.  I find the single line compounded form harder to grok.